Network Indexer Dev Tools Page (#471)
protocol/launchpad: content/en/curriculum/dev-tools/network-indexer/index.md

---
title: "Network Indexer (IPNI)"
description: "Indexing and enabling Faster retrieval for IPFS and Filecoin"
draft: false
menu:
  curriculum:
    parent: "curriculum-devtools"
weight: 80
category: lecture
level:
- shallow
- deep
---

## <Dev Tool> Introduction
The [IPNI network indexer](https://github.com/ipni) is a project created and maintained by Protocol Labs that is designed to index the data on the Filecoin and IPFS networks, and work alongside the Distributed Hash Table (DHT) employed by both the IPFS and Filecoin networks to enable the fast and efficient retrieval of content-addressed data.

## Why Use Indexed Content Routing?
Protocol Labs' Network Indexer (IPNI) enables any user or developer to query both the IPFS and Filecoin public nodes to find content-addresses data using the [CID](../ipfs/content-addressing/#content-addressing).

Since IPFS and Filecoin use different protocols to retrieve data (IPFS uses Bitswap, and Filecion uses Graphsync), there is a need for a solution that makes it possible for anyone who would want to retrieve and use that data to locate and understand which protocol they can use to retrieve that data and use it in other applications.

> With versions of kubo 16 and above, a feature was implemented that makes it possible to use HTTP to send and receive information about IP addresses of peers that have certain content (CIDs) and publish IPNS records.

Reviewer: Kubo 16
```suggestion
Kubo 16 introduced Delegated Routing protocol that allows a node to publish its CIDs to alternative routing systems alongside the DHT and be used for subsequent content-addressed data lookups. 
```

The Indexer also makes it possible, in a way not achieved before, for people to retrieve data from the Filecoin network, as there is a lot of data stored across Filecoin nodes, but  no way to search and retrieve that data searching either by the CID of multihash of that data.

Though the DHT is an amazing, distributed way to advertise and discover content, the IPFS indexer adds another layer that can give you other options for retrieving files for your applicaitons. With an open protocol that anyone can use, as seen here in the [go-delgated-routing](https://github.com/ipfs/go-delegated-routing) library on their server.

> to enable faster discovery and routing, along with load balancing and an alternative market where routing is done, besides the DHT.

Reviewer: ```suggestion
With an open protocol that anyone can use, as seen here in the [go-delgated-routing](https://github.com/ipfs/go-delegated-routing) library on their server.
to enable faster discovery and routing, along with load balancing and an alternative market where routing is done, besides the DHT
```
IPNI is an open protocol that anyone can use to quickly find content-addressed data on IPFS or Filecoin, to index the data that they host or to run an indexer of their own. IPNI enables faster discovery and routing, along with load balancing and an alternative market where routing is done, besides the DHT. IPNI integrates with IPFS through the [go-delegated-routing](https://github.com/ipfs/go-delegated-routing) and [index-provider](https://github.com/ipni/index-provider) libraries.

Laern more about IPFS Indexer, the motivations behind it, and potential use cases, see the video **[State of Content Routing| IPFS Camp Lisbon 2022](https://www.youtube.com/watch?v=DLCTEXbF1Es)**

## Indexer Implementations

> Since Filecoin and IPFS store data on separate networks, using different methods of data transfer, there are who different ways that the information about the CIDs are communicated to IPNI.

Reviewer: ```suggestion
IPNI builds its index by processing (Advertisements)[https://github.com/ipni/storetheindex/blob/main/api/v0/ingest/schema/schema.ipldsch#L40]. The Advertisement construct allows a Storage Provider to offer their CIDs to IPNI, which would make these CIDs available for fast lookups once the Advertisement has been processed. Since Filecoin and IPFS store data on separate networks, using different methods of data transfer, there are who different ways that the information about the CIDs are communicated to IPNI.
```

> ### Filecoin Indexers

Reviewer: Title `Filecoin Indexers` might give an impression that IPFS and Filecoin are indexer by physically different systems while in reality it's not true. Maybe rename it to "Advertising from Filecoin", "Advertising from IPFS", or something along these lines?

> Storage providers such as [Estuary](https://estuary.tech/) who store on Filecion can publish their content to the indexer using [index-provider](https://github.com/ipni/index-provider) library. When a deal is made and data is published to Filecoin, the would also [advertise](https://github.com/ipni/storetheindex/blob/main/api/v0/ingest/schema/schema.ipldsch#L40) to IPNI. This means that the CIDs and the information about the data in that CID is published on libp2p to a topic that IPNI follows.

Reviewer: ```suggestion
An update is made available on a libp2p to a topic that IPNI follows and then reaches out to the Storage Providers for an update when sees it. 
```

> Storage providers such as [Estuary](https://estuary.tech/) who store on Filecion can publish their content to the indexer using [index-provider](https://github.com/ipni/index-provider) library. When a deal is made and data is published to Filecoin, the would also [advertise](https://github.com/ipni/storetheindex/blob/main/api/v0/ingest/schema/schema.ipldsch#L40) to IPNI. This means that the CIDs and the information about the data in that CID is published on libp2p to a topic that IPNI follows.

Reviewer: ```suggestion
Storage providers such as [Estuary](https://estuary.tech/) who store on Filecion can publish their content to the indexer using [index-provider](https://github.com/ipni/index-provider) library. When a deal is made and data is published to Filecoin, the would also [advertise](https://github.com/ipni/storetheindex/blob/main/api/v0/ingest/schema/schema.ipldsch#L40) to IPNI. An update is made available on a libp2p to a topic that IPNI follows and then reaches out to the Storage Providers for an update when sees it. 
```

### IPFS Indexers

> When data is added to the IPFS network via kubo version 16 and above, there is an API used with the Reframe or [go-delegated-routing](https://github.com/ipfs/go-delegated-routing) that is used to send data about those CIDs in IPNI. Anyone who published to ipfs using Kubo can configure their node to advertise CIDs into both the DHT and IPNI.

Reviewer: ```suggestion
When data is added to the IPFS network via kubo version 16 and above, there is an API used with the Reframe or [go-delegated-routing](https://github.com/ipfs/go-delegated-routing) that is used to send data about those CIDs to IPNI. Anyone who published to ipfs using Kubo can configure their node to advertise CIDs into both the DHT and IPNI.
```

#### Reframe

> [Reframe](https://github.com/ipfs/go-delegated-routing) is a Request-response Protocol (RPC) used by IPFS kubo 16+ that kicks off a procedure or subroutine to execute on the IPFS network that stores the data in an index, and allows others to retrieve information related to that CID.

Reviewer: ```suggestion
[Reframe](https://github.com/ipfs/go-delegated-routing) is a Request-response Protocol (RPC) used by IPFS kubo 16+ that allows the node to publish its own CIDs as well as to look up other CIDs in an alternative routing system such as IPNI alongside the DHT. That doesn't have to be either / or choice - content can be looked up in both and whichever is faster wins. 
```

With versions of kubo 16+, the Refame RPC is integrated as a feature
* Reframe adds an additional way to discover peers, content, and IPNS records
* With Reframe, you can configure your IPFS kubo node to publish a snapshot of all of the CIDs on your node at whatever frequency you would like

> * Reframe uses HTTP transport to store this information on an indexer node

Reviewer: Reframe doesn't speak to indexers directly. Instead it relies on someone translating its messages into the IPNI protocol. Might be better to rephrase it to "Reframe is built on top of HTTP."

Author: Updated the paragraph

> * Reframe is an alternative to the DHT

Reviewer: The idea behind reframe was to democratize content routing and enable a market where alternative content routing systems can emerge. We shouldn't be advertising it as a replacement for DHT - rather something that work alongside DHT and be better for some usecases.

With an open protocol that anyone can use, as seen here in the [go-delgated-routing](https://github.com/ipfs/go-delegated-routing) library on their server.
to enable faster discovery and routing, along with load balancing and an alternative market where routing is done, besides the DHT

Laern more about IPFS Indexer, the motivations behind it, and potential use cases, see the video **[State of Content Routing| IPFS Camp Lisbon 2022](https://www.youtube.com/watch?v=DLCTEXbF1Es)**

## Indexer Implementations
IPNI builds its index by processing (Advertisements)[https://github.com/ipni/storetheindex/blob/main/api/v0/ingest/schema/schema.ipldsch#L40]. The Advertisement construct allows a Storage Provider to offer their CIDs to IPNI, which would make these CIDs available for fast lookups once the Advertisement has been processed. Since Filecoin and IPFS store data on separate networks, using different methods of data transfer, there are who different ways that the information about the CIDs are communicated to IPNI.

> ### Filecoin Indexers

Author: ```suggestion
### Advertising Filecoin CIDs
```

Storage providers such as [Estuary](https://estuary.tech/) who store on Filecion can publish their content to the indexer using [index-provider](https://github.com/ipni/index-provider) library. When a deal is made and data is published to Filecoin, the would also [advertise](https://github.com/ipni/storetheindex/blob/main/api/v0/ingest/schema/schema.ipldsch#L40) to IPNI. This means that the CIDs and the information about the data in that CID is published on libp2p to a topic that IPNI follows.

> ### IPFS Indexers

Author: ```suggestion
### Advertising IPFS CIDs
```

The Indexer also makes it possible, in a way not achieved before, for people to retrieve data from the Filecoin network, as there is a lot of data stored across Filecoin nodes, but  no way to search and retrieve that data searching either by the CID of multihash of that data.

Though the DHT is an amazing, distributed way to advertise and discover content, the IPFS indexer adds another layer that can give you other options for retrieving files for your applicaitons. With an open protocol that anyone can use, as seen here in the [go-delgated-routing](https://github.com/ipfs/go-delegated-routing) library on their server.
With an open protocol that anyone can use, as seen here in the [go-delgated-routing](https://github.com/ipfs/go-delegated-routing) library on their server.
to enable faster discovery and routing, along with load balancing and an alternative market where routing is done, besides the DHT

Laern more about IPFS Indexer, the motivations behind it, and potential use cases, see the video **[State of Content Routing| IPFS Camp Lisbon 2022](https://www.youtube.com/watch?v=DLCTEXbF1Es)**

## Indexer Implementations

> IPNI builds its index by processing (Advertisements)[https://github.com/ipni/storetheindex/blob/main/api/v0/ingest/schema/schema.ipldsch#L40]. The Advertisement construct allows a Storage Provider to offer their CIDs to IPNI, which would make these CIDs available for fast lookups once the Advertisement has been processed. Since Filecoin and IPFS store data on separate networks, using different methods of data transfer, there are who different ways that the information about the CIDs are communicated to IPNI.

Author: ```suggestion
IPNI builds its index by processing [Advertisements](https://github.com/ipni/storetheindex/blob/main/api/v0/ingest/schema/schema.ipldsch#L40). The Advertisement construct allows a Storage Provider to offer their CIDs to IPNI, which would make these CIDs available for fast lookups once the Advertisement has been processed. Since Filecoin and IPFS store data on separate networks, using different methods of data transfer, there are who different ways that the information about the CIDs are communicated to IPNI.
```

### Advertising Filecoin CIDs
Storage providers such as [Estuary](https://estuary.tech/) who store on Filecion can publish their content to the indexer using [index-provider](https://github.com/ipni/index-provider) library. When a deal is made and data is published to Filecoin, the would also [advertise](https://github.com/ipni/storetheindex/blob/main/api/v0/ingest/schema/schema.ipldsch#L40) to IPNI. An update is made available on a libp2p to a topic that IPNI follows and then reaches out to the Storage Providers for an update when sees it. 

### Advertising IPFS CIDs
When data is added to the IPFS network via kubo version 16 and above, there is an API used with the Reframe or [go-delegated-routing](https://github.com/ipfs/go-delegated-routing) that is used to send data about those CIDs to IPNI. Anyone who published to ipfs using Kubo can configure their node to advertise CIDs into both the DHT and IPNI.

#### Reframe

> [Reframe](https://github.com/ipfs/go-delegated-routing) is a Request-response Protocol (RPC) used by IPFS Kubo 16+ that allows the node to publish its own CIDs as well as to look up other CIDs in an alternative routing system such as IPNI alongside the DHT. That doesn't have to be either / or choice - content can be looked up or published to both in the same time. 

Author: ```suggestion
[Reframe](https://github.com/ipfs/go-delegated-routing) is a Request-response Protocol (RPC) used by IPFS Kubo 16+ that enables the translation of CIDs using the HTTP protocol to IPNI.
```

Kubo doesn't know how to speak to IPNI by itself. Instead it relies on someone translating reframe messages into the IPNI protocol. That is done to enable alternative routing systems market without being bound to the DHT or IPNI protocols. 

That job is done by Index Provider process that can be run alongside a Kubo node as a bridge to IPNI. 
With versions of kubo 16+, the Refame RPC is integrated as a feature

> * Reframe adds an additional way to discover peers, content, and IPNS records

Author: ```suggestion
* Reframe adds an additional way to discover peers, content, and IPNS records alongside the DHT
```

  * Make sure that `GOPATH/bin` is on your `PATH`

```
$ export GOPATH=$HOME/go
$ export PATH=$PATH:$GOPATH/bin
```
* kubo (IPFS go) version 16 or above installed
  * IPFS install [Tutorial](https://curriculum.pl-launchpad.io/tutorials/ipfs-intro/setup/)
  * [Docs Instructions](http://docs.ipfs.tech.ipns.localhost:8080/install/command-line/#install-official-binary-distributions)
  * [Distributions](https://dist.ipfs.tech/#kubo)
* [IPFS companion](https://chrome.google.com/webstore/detail/ipfs-companion/nibjojkomfdiaoajekhjakgkdhaomnch?hl=en) browser extension or [Brave browser](https://brave.com/download/) to view websites published on IPFS
- Install a [JSON formatter extensio](https://chrome.google.com/webstore/detail/json-formatter/bcjindcccaagfpapjjmafapmmgkkhgoa?hl=en)n for your web browser

## Tutorial Instructions
You can see the video of the workshop called **[Indexed Content Routing](https://www.youtube.com/watch?v=aN7fGturjzA&t=121s)** from IPFS Camp Lisbon 2022. The original instructions are at [https://github.com/ischasny/ipfs-camp-routing](https://github.com/ischasny/ipfs-camp-routing).

### Explore IPNI Records
First, lets take a look at an example IPNI record.

> * Check out the Filecion Saturn website, [located at the CID bafybeia57mwbxw3csprt72a6bd6o4uedazn3vo6tv64xken6fgmaxtiugy](https://bafybeia57mwbxw3csprt72a6bd6o4uedazn3vo6tv64xken6fgmaxtiugy.ipfs.dweb.l).

Author: ```suggestion
* Check out the Filecion Saturn website, located at the CID `bafybeia57mwbxw3csprt72a6bd6o4uedazn3vo6tv64xken6fgmaxtiugy`.
```

![Peer ID](ipfspeerid.png)

### Set Up an index-provider Node

See [https://github.com/ipni/index-provider#install](https://github.com/filecoin-project/index-provider#install) for full instructions on run a standalone provider daemon instance.

Install index-provider:

```
$ go install github.com/filecoin-project/index-provider/cmd/provider@v0.9.0
```

Initialise the index-provider repository and configuration:

```
$ provider init
```

Add a direct http announcement (optional) and Reframe configuration:

```
$ vim ~/.index-provider/config
```

Add the following configuration block replacing the ProviderID with the kubo node peer ID grabbed at the previous step:

```
 "DirectAnnounce": {
    "URLs": [
      "http://127.0.0.1:3001"
    ]
  },
  "Reframe": {
    "ListenMultiaddr": "/ip4/127.0.0.1/tcp/50617",
    "ChunkSize": 1,
    "SnapshotSize": 100,
    "ProviderID": "KUBO_NODE_PEER_ID",
    "Addrs": [
      "/ip4/0.0.0.0/tcp/4001",
      "/ip6/::/tcp/4001",
      "/ip4/0.0.0.0/udp/4001/quic",
      "/ip6/::/udp/4001/quic"
    ]
  }
```

Start the index-provider:

```
$ provider daemon
```

### See It In Action

> Make sure `ipfs daemon` is running and open up a WebUI at [http://127.0.0.1:5001/webui](http://127.0.0.1:5001/webui) and upload a file.

Author: ```suggestion
Make sure `ipfs daemon` is running and open up a WebUI at [http://127.0.0.1:5001/webui](http://127.0.0.1:5001/webui) and upload a file.

You can also add a file using the `ipfs add path/to/local/file` command and take note of the CID output in the terminal
```

---
title: "Network Indexer (IPNI)"
description: "Indexing and enabling Faster retrieval for IPFS and Filecoin"
draft: false
menu:
  curriculum:
    parent: "curriculum-devtools"
weight: 80
category: lecture
level:
- shallow
- deep
---

## <Dev Tool> Introduction
The [IPNI network indexer](https://github.com/ipni) is a project created and maintained by Protocol Labs that is designed to index the data on the Filecoin and IPFS networks, and work alongside the Distributed Hash Table (DHT) employed by both the IPFS and Filecoin networks to enable the fast and efficient retrieval of content-addressed data.

## Why Use Indexed Content Routing?

> Protocol Labs' Network Indexer (IPNI) enables any user or developer to query both the IPFS and Filecoin public nodes to find content-addresses data using the [CID](../ipfs/content-addressing/#content-addressing).

Author: ```suggestion
Protocol Labs' Network Indexer (IPNI) enables any user or developer to query both the IPFS and Filecoin public nodes to find content-addresses data using the [CID](../ipfs/content-addressing/#content-addressing). The cid.contact project is the largest IPNI deployment to date, managed by PL. However there are other deployments too. cid.contact can be explored via a [UI](https://cid.contact/) or via [API](https://cid.contact/cid/bafybeia57mwbxw3csprt72a6bd6o4uedazn3vo6tv64xken6fgmaxtiugy). 
```

Author: @ischasny first part of your comment added here

> Since IPFS and Filecoin use different protocols to retrieve data (IPFS uses Bitswap, and Filecion uses Graphsync), there is a need for a solution that makes it possible for anyone who would want to retrieve and use that data to locate and understand which protocol they can use to retrieve that data and use it in other applications.

Reviewer: Maybe add "More information about IPNI can be found here": https://github.com/ipni/specs/blob/main/IPNI.md

Reviewer: Would be good to mention that cid.contain is the largest IPNI deployment to date, managed by PL. However there are other deployments too. cid.contact can be explored via a [UI](https://cid.contact) or via (API)[https://cid.contact/cid/bafybeia57mwbxw3csprt72a6bd6o4uedazn3vo6tv64xken6fgmaxtiugy]. The API call above returns a `json` object that says where and how the data represented with a CID `bafybeia57mwbxw3csprt72a6bd6o4uedazn3vo6tv64xken6fgmaxtiugy` can be fetched from.

Reviewer: We can also mention some stats that at the moment of writing cid.contact has 1.3 trillion CIDs indexed that is spread across >400 [providers](https://cid.contact/providers) and ingests over 10 billion new CIDs weekly.

Author: I want to avoid adding numbers & istats that are changing -- it would mean we would have to go in and update this document almost weekly, but we can mention general things like 'The largest at this time' which won't change. Is there a place these stats are published that we can point to instead?

Reviewer: Not at the moment. That's ok too.

Kubo 16 introduced Delegated Routing protocol that allows a node to publish its CIDs to alternative routing systems alongside the DHT and be used for subsequent content-addressed data lookups.

> IPNI also makes it possible, in a way not achieved before, for people to retrieve data from the Filecoin network, as there is a lot of data stored across Filecoin nodes, but  no way to search and retrieve that data searching either by the CID of multihash of that data.

Author: ```suggestion
IPNI also makes it possible, in a way not achieved before, for people to retrieve data from the Filecoin network, as there is a lot of data stored across Filecoin nodes, but  no way to search and retrieve that data searching either by the CID of multihash of that data. The API calls return json objects that contain information about where and how the data represented with a CID such as `bafybeia57mwbxw3csprt72a6bd6o4uedazn3vo6tv64xken6fgmaxtiugy` can be fetched from.
```

> Though the DHT is an amazing, distributed way to advertise and discover content, the IPFS indexer adds another layer that can give you other options for retrieving files for your applicaitons. With an open protocol that anyone can use, as seen here in the [go-delgated-routing](https://github.com/ipfs/go-delegated-routing) library on their server.

Reviewer: ```suggestion
Though the DHT is an amazing, distributed way to advertise and discover content, IPNI adds another layer that can give you other options for retrieving files for your applications. With an open protocol that anyone can use, as seen here in the [go-delgated-routing](https://github.com/ipfs/go-delegated-routing) library on their server.
```

With an open protocol that anyone can use on their server, as seen here in the [go-delgated-routing](https://github.com/ipfs/go-delegated-routing) library.
to enable faster discovery and routing, along with load balancing and an alternative market where routing is done, besides the DHT

> Laern more about IPFS Indexer, the motivations behind it, and potential use cases, see the video **[State of Content Routing| IPFS Camp Lisbon 2022](https://www.youtube.com/watch?v=DLCTEXbF1Es)**

Reviewer: ```suggestion
Laern more about IPNI, the motivations behind it, and potential use cases, see the video **[State of Content Routing| IPFS Camp Lisbon 2022](https://www.youtube.com/watch?v=DLCTEXbF1Es)**
```

#### Reframe
[Reframe](https://github.com/ipfs/go-delegated-routing) is a Request-response Protocol (RPC) used by IPFS Kubo 16+ that allows the node to publish its own CIDs as well as to look up other CIDs in an alternative routing system such as IPNI alongside the DHT. That doesn't have to be either / or choice - content can be looked up or published to both in the same time.

Kubo doesn't know how to speak to IPNI by itself. Instead it relies on someone translating reframe messages into the IPNI protocol. That is done to enable alternative routing systems market without being bound to the DHT or IPNI protocols.

That job is done by Index Provider process that can be run alongside a Kubo node as a bridge to IPNI.
With versions of kubo 16+, the Refame RPC is integrated as a feature
* Reframe adds an additional way to discover peers, content, and IPNS records
* With Reframe, you can configure your IPFS kubo node to publish a snapshot of all of the CIDs on your node at whatever frequency you would like
* Reframe uses HTTP transport to store this information on an indexer node

> * Reframe is an alternative to the DHT

Reviewer: ```suggestion
* Reframe makes alternative routing systems other than DHT possible
```

> * Content can be published to both the IPFS DHT and Reframe with a tool called parallel

Author: ```suggestion
* Content can be published to both the IPFS DHT and Reframe with a tool called ParallelHelper
```

> * Content can be published to both the IPFS DHT and Reframe with a tool called parallel

Reviewer: This block ends abruptly, probably a typo?

Author: Changed it to say ParallelHelper

  Reframe was created from a [kubo spec](https://github.com/ipfs/specs/blob/main/reframe/REFRAME_PROTOCOL.md) (the IPFS go implementation) that allows IPFS nodes to advertise their content to other systems besides the DHT.

## Tutorial: IPNI

> In this simple tutorial, we are going to configure and run an IPFS node to communicate data to the indexer and take a look at what that data looks like.

Reviewer: Maybe add:

We are going to setup three things:
1. Kubo IPFS node
2. `index-provider` that is going to be used as a bridge between IPNI and IPFS
3. Network indexer that the IPFS node is going to advertise its contents to.

Author: See below

#### Workshop: Indexed Content Routing| IPFS Camp Lisbon 2022

> See how to set up a local indexer and index-provider in this workshop from IPFS Camp 2022.

Author: ```suggestion
See how to set up a local indexer and index-provider in this workshop from IPFS Camp 2022. In this tutorial you will: 
* Modify the config for your kubo IPFS node
* Set up a local network indexer that the IPFS node is going to advertise its contents to
* Set up index-provider that is going to be used as a bridge between IPNI and IPFS

```

## Tutorial Instructions
You can see the video of the workshop called **[Indexed Content Routing](https://www.youtube.com/watch?v=aN7fGturjzA&t=121s)** from IPFS Camp Lisbon 2022. The original instructions are at [https://github.com/ischasny/ipfs-camp-routing](https://github.com/ischasny/ipfs-camp-routing).

### Explore IPNI Records
First, lets take a look at an example IPNI record.
* Check out the [Filecion Saturn website](https://strn.network/), published with the CID `bafybeia57mwbxw3csprt72a6bd6o4uedazn3vo6tv64xken6fgmaxtiugy`.
* Now, look up this CID on [CID contact](https://cid.contact/)
  ![CID Contact](cid-contact2.png)
* You can also see that same information by constructing the cid.contact url like so:
```
https://cid.contact/cid/<CID>
```

> This one is located at [https://cid.contact/cid/bafybeia57mwbxw3csprt72a6bd6o4uedazn3vo6tv64xken6fgmaxtiugy](https://cid.contact/cid/bafybeia57mwbxw3csprt72a6bd6o4uedazn3vo6tv64xken6fgmaxtiugy)

* Take a look at the providers who are broadcasting CIDs to the network at [https://cid.contact/providers](https://cid.contact/providers)

### Set up a Local Indexer

> Installing and running the indexer on your local machine makes it possible for your kubo node to communicate with the indexer, by creating the advertisements in a format the indexer can ingest.

Author: ```suggestion
Installing and running the indexer on your local machine makes it possible for your kubo node to communicate with the indexer, by creating the advertisements in a format the indexer can ingest.

In this tutorial we are going to run an indexer locally, however it's also possible to advertise your data to cid.contact directly. Running a local node makes it possible for cid.contact to dial back into your node to fetch a chain of advertisements. This is possible without extra firewall configurations because of the provider-node setup in the next sections.
```

You can see the documentation, CLI commands, and thorough instructions at [https://github.com/filecoin-project/storetheindex/#install](https://github.com/filecoin-project/storetheindex/#install).

Install storetheindex:

```
$ go install github.com/filecoin-project/storetheindex@v0.4.28
```

Initialize the storetheindex repository and configuration:

```
$ storetheindex init
```

Start the indexer

```
$ storetheindex daemon
```

> 

Reviewer: Might be worth to clarify:

In this tutorial we are going to run an indexer locally, however it's also possible to advertise your data to cid.contact directly. The reason why we are not going to do that now - is because cid.contact needs to be able to dial back into your node to fetch a chain of advertisements. However, that might be problematic to do behind home routers and would require some additional firewall configuration.

Author: Added to top of this part of the conversation -- I thought that index-provider solved problems posed by home firewall configurations?

```json
...
    "Reframe": {
      "ListenMultiaddr": "/ip4/127.0.0.1/tcp/50617",
      "ChunkSize": 1,
      "SnapshotSize": 100,
      "ProviderID": "12D3KooWBEQR33uW1T6axtpspeyJo6m8Sy531fQ8mzQbW1JBnWxJ",
      "Addrs": [
        "/ip4/0.0.0.0/tcp/4001",
        "/ip6/::/tcp/4001",
        "/ip4/0.0.0.0/udp/4001/quic",
        "/ip6/::/udp/4001/quic"
      ]
    }
}
```

Start the index-provider:

```
$ provider daemon
```

### See It In Action

Make sure `ipfs daemon`, the `provider daemon`, and `storetheindex daemon` are all running. You should also have another terminal window open to add & check to see if store-the-index is working.

![IIPNI Commands](ipni-commands.png)

#### Add a File
Open up a WebUI at [http://127.0.0.1:5001/webui](http://127.0.0.1:5001/webui) and upload a file, then copy the CID (right click on the file).

You can also add a file using the `ipfs add path/to/local/file` command and take note of the CID output in the terminal

Once upoloaded you should see logs start rolling in `storetheindex` and `indexprovider` command lines.

#### Check the Index
Open a new terminal window and check whether your cid got indexed by executing

```

> $ storetheindex find --cid <CID>

Author: ```suggestion
storetheindex find --cid <CID>
```
